doc/platform/nlapi/netlink_api_kernel_programming.md

```diff
@@ -0,0 +1,223 @@
+
+
+### 3.1.1    Performance comparison
+
+In order to benchmark the performance improvement, SwSS Vlan manager was considered for measuring the performance difference between fork and API based execution. Significant performance improvement in SwSS Vlan configuration manager kernel programming was observed.  Measured data is as shown in table below. 
+
+Total time taken at SwSS: Shell fork calls vs Netlink API calls  
+
+| **Config Operation** | **Fork Calls** | **Netlink API Calls** | **Performance gain** |  
+|:---------------------|:--------------:|:---------------------:|:---------------------------:|  
+| One VLAN config when system is loaded     | 115 msec    | 08 msec    | 14 times |  
+| 4000 VLAN configuration                   | 305 seconds | 21 seconds | 14 times |  
+| 4000 VLAN config using VLAN range command | 309 seconds | 13 seconds | 23 times |  
+| 4000 VLAN config with system restart      | 406 seconds | 15 seconds | 27 times |  
+| 4000 VLAN config  with config file load   | 404 seconds | 17 seconds | 23 times |  
+
+### Table 2: Performance comparision  
+
+Command set used for above performance measurement.  
+
+       One Vlan config: 
+           config vlan add 9  
+           config vlan member add 9 Ethernet20  
+           config vlan member add 9 PortChannel005  
+       4000 Vlan config:   
+            config vlan range add 10 4009   
+            config vlan member range add 10 4009 Ethernet20  
+            config vlan member range add 10 4009 PortChannel005     
+
+
+Netlink API library will be placed in sonic-swss path with directory named 'nlapi'.
+Following APIs will be supported in current release
+
+      
+      int nlapi_req_ip_link_add_dev(const char *dev, const char *type, const char *admin)
+      int nlapi_req_ip_link_add_vrf(const char *vrf_name, unsigned int vrf_table)
+      int nlapi_req_ip_link_add_vxlan(const char *vx_lan_if)
+      int nlapi_req_ip_link_add_vxlan_id(const char *dev, int vxlan_id, const char *lcl_addr, int dest_port)
+      int nlapi_req_ip_link_add_link_vlan(const char *dev, unsigned int vlan_id, const char *admin, const char *mac_addr)
+      int nlapi_req_ip_link_set_admin(const char *dev, const char *admin)
+      int nlapi_req_ip_link_set_mtu(const char *dev, unsigned int mtu)
+      int nlapi_req_ip_link_set_vlan(unsigned int vlan_id, const char *admin)
+      int nlapi_req_ip_link_set_master(const char *dev, const char *master, const char *admin)
+      int nlapi_req_ip_link_set_filtering(const char *dev, const char *type, unsigned int filter_id)
+      int nlapi_req_ip_link_del_dev(const char *dev)
+      int nlapi_req_ip_link_list(const char *dev, unsigned int display, int *count)
+      int nlapi_req_ip_link_list_type(const char *type, int display, int *count)
+      int nlapi_req_ip_link_list_master_ifs(const char *dev, int display, int *count)
+      int nlapi_req_ip_address_cmd(const char *action, int version, const char *prefix, const char *dev)
```

Comment:
Can we use enum for add/del actions? This can be used for other APIs as well and have common API for add/del based on the enum type.